source/compiler-core/slang-struct-tag-converter.h

```diff
@@ -0,0 +1,153 @@
+#ifndef SLANG_COMPILER_CORE_STRUCT_TAG_CONVERTER_H
+#define SLANG_COMPILER_CORE_STRUCT_TAG_CONVERTER_H
+
+#include "slang-struct-tag-system.h"
+#include "slang-diagnostic-sink.h"
+
+namespace Slang {
+
+class StructTagConverterBase
```

Comment:
Maybe I'm missing something but this seems like a lot of infrastructure compared to the simpler thing of just having each relevant function do `switch(tag) { ... }` and either handle each allowed case directly (there's only one right now!) or convert the data for each case into a common representation.

For handling extensions, it seems like we'd often want code along the lines of:

```
// one local per recognized extension and version thereof
ExtA* a = nullptr;
ExtB* b = nullptr;
for(auto extPtr : primary->exts) {
    switch(extPtr->tag) {
        case A: a = (A*) extPtr; break;
        ...
    }
}
// later code can easily do:
if(a) { /* a-specific stuff */ }
```

I'm beginning to see that we maybe had very different ideas of how much of an underlying *system* needed to exist behind this. In my mind the tags are just a glorified `enum` to indicate which of a list of `struct` types was passed in, and the logic for actually handling those cases would be simplest and fastest if it is ad hoc...

Reply:
This is nice in so far as allows for programmatic transformations between versions. The PR solution provides mechanical transformations around semantic versioning (for now at least).

On the other hand the mechanism above implies that all the versions are available, and that any handling is ad-hoc. 

Whilst the mechanical semantic versioning is more limited, once it exists - it does not require keeping previous versions around, or ad-hoc processing or error handling. You do it once and you are done.